multidim-interop/go/v0.24/main.go

```diff
@@ -0,0 +1,284 @@
+package main
```

Comment:
We're replacing the shims with multiple  mains:

We initially implemented "one main, multiple shims" to address the explosions of versions (https://github.com/libp2p/test-plans/pull/23#pullrequestreview-960603753).

I'm not sure what's motivating this change; I think this is a regression:
Changing the test specs, like adding benchmarking or new parameters will become exponentially difficult (N implementations x M main files).

Reply:
I should talk with Marten. But I don't really think this is a problem in this specific test case. We want to have implementation A ping implementation B and nothing else. I bootstrapped off the existing ping cases because it was convenient.

In fact, one version of this branch had me define patches that can be applied to other versions to do something similar to the "one main, multiple shims" idea, but I realized it was needlessly complex. Once a version is out and published there shouldn't be much changes to the implementation. Users using v0.22, for example, aren't going to update their implementation to interact with newer versions. Why are we assuming we are going to go back and change the implementation of this test case in v0.22?

If we want to add benchmarking or new parameters that's a new test. Any shim we design before this new test isn't guaranteed to avoid the N x M problem.

Reply:
Once a version is out and published there shouldn't be much changes to the implementation. Users using v0.22, for example, aren't going to update their implementation to interact with newer versions. Why are we assuming we are going to go back and change the implementation of this test case in v0.22?
> 
> If we want to add benchmarking or new parameters that's a new test. Any shim we design before this new test isn't guaranteed to avoid the N x M problem.

I did the exact same thing for Rust with the same reasoning :)

Eventually I hope we can track these tests in Git by having them next to our source code which means there is only just one to maintain and it moves along with the APIs and features as we go, just the published binary remains as a test artifact and gets run by testground when requested.

Reply:
Thanks for sharing detailed answers!

> Users using v0.22, for example, aren't going to update their implementation to interact with newer versions. Why are we assuming we are going to go back and change the implementation of this test case in v0.22?

> there is only just one to maintain and it moves along with the APIs and features as we go,

> If we want to add benchmarking or new parameters that's a new test. Any shim we design before this new test isn't guaranteed to avoid the N x M problem.

Got it, I see the nuance.

I was starting from the constraints: we want to add and update tests often. New tests should cover _most_ known versions, even old ones, as easily as possible. The team wants to write the minimum amount of test code.

The new approach was itching because it seems we end up with a lot of duplicated code and we'll have to revisit all these mains (and later release branches) whenever the spec changes. Whereas with shims, we can update one `main.go` (the spec) and it applies to every known version.

That makes sense now, we're less worried about updating test specs and reusing code because we start from the assumptions: test specs won't change often, and most of these specs will be written for the current and future versions.

Thanks for the clarification.

Reply:
I have mixed feelings on this one. I see the argument that existing test plans will not be updated frequently, although I do expect us to make changes to the ping test plan, 300 LOC are just waaaay too complex for a simple ping test. For example, I'd like to see the network configuration being moved outside of the test plan - this doesn't need to be a runtime configuration. We can also reduce complexity by making this a test that only 2 hosts are participating in explicitly.

I imagine that future tests will share a fair bit of code with the ping test, e.g. constructing a host with a certain transport configuration. I can see us introducing helper functions _per version_ (and not _cross_ version, as a shim would do) to avoid code duplication.

Just my 2 cents, definitely not a blocker for this PR.

Reply:
>  I can see us introducing helper functions per version (and not cross version, as a shim would do) to avoid code duplication.

:+1:

As a data point, I'm trying to fix a [race condition](https://github.com/libp2p/test-plans/pull/85/commits/102f01fcec39ce6c634bbf961f9b8a102e74a4dd#diff-324e6c18346c026bb1db7c265dfd87cdcca67e99f52a317966321453187e1505R223) in our existing Go ping test-plan and using the compat shim is way harder than doing it once per version. In this branch I did the once per version and was done in <5 minutes. With the compat shim, I can't even run `go build` to see if I did the right thing.